### PR TITLE
Taxis algorithm

### PR DESCRIPTION
**Pull request by vagechirkov**
Phototaxis algorithm refactoring and discussion

### CODE REVIEW

File: abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py [hunk 1]
```diff
@@ -36,3 +36,34 @@ def F_reloc_LR(vel_now, V_now, v_desired=None, theta_max=None):
...
+        new_sign = np.sign(prev_theta) if prev_theta != 0 else 1
+    # negative means we need to turn the other direction
+    else:
+        if taxis_dir is None:
+            # new sign is a flipped sign of the previous theta
+            new_sign = sign_diff * np.sign(prev_theta)
+        else:
+            new_sign = taxis_dir
+    # change theta proportional to the difference in meter values
+    new_theta = phototaxis_theta_step * new_sign * np.abs(diff)
```
Comment (vagechirkov) [inline abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py]: @mezdahun what is your intuition to change the theta value proportionally to the raw value of the meter (detector)?

File: abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py [hunk 2]
```diff
@@ -36,3 +36,32 @@ def F_reloc_LR(vel_now, V_now, v_desired=None, theta_max=None):
...
+    # negative means we need to turn the other direction
+    else:
+        if taxis_dir is None:
+            # new sign is a flipped sign of the previous theta
+            new_sign = sign_diff * np.sign(prev_theta)
+        else:
+            new_sign = taxis_dir
+    # change theta proportionally to the meter values
+    # the larger the meter value, the smaller the turning angle
+    new_theta = phototaxis_theta_step * new_sign * (1 - meter)
```
Comment (vagechirkov) [inline abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py]: @mezdahun what is your intuition to change the theta value proportionally to the raw value of the meter (detector)? I think it makes more sense to change it proportionally to 1 - meter. What do you think?

Comment (mezdahun) [reply abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py]: As the agents are continuously forced to move with a given speed, when agents are close to the center they have to turn relatively faster to oscillate around the patch center

**Comment by mezdahun** [comment]
Hey there, in commits https://github.com/scioip34/ABM/pull/49/commits/36e85df268e3a28aa6cd2c4508685a802c2d1426 and https://github.com/scioip34/ABM/pull/49/commits/da420a9c467d7f4b87d1d2b083f708a6ef0f1cb4 I have fixed the implementation of the phototaxis algorithm by clarifying what `taxis_dir` was and passed it back to the agent class.

As further note: the current algorithm consists of a snake-like movement with rotation to a given direction until the meter value of an agent is increasing. In these cases the `taxis_dir` is None and we act according to the meter value. As soon as the meter value drops we assume that our change of direction (especially it's sign, i.e. left or right) was non-optimal, so we initiate a directional switch. This we do by setting the `taxis_dir` to +-1. The agent will then in the next steps act according to this requested direction until the meter value increases again.

**Comment by mezdahun** [comment]
In commit https://github.com/scioip34/ABM/pull/49/commits/67ad42b908c7bfb5a33d0a845ab568eb9c984ccf I changed the test for phototaxis to represent the fixed implementation with direct proportional turning to meter value and incorporated return of directional change (`new_taxis_dir`).

File: abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py (hunk 2 above)
Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py]: That makes sense! Thank you for the explanation!

**Comment by vagechirkov** [comment]
I'm wondering if this is the expected behavior (see attached video). In general, I find phototaxis a bit inefficient. I'm curious if this also applies to humans and animals in a similar environment. I think we're good at bringing this PR together, I have no comments on the implementation itself 🙃

taxis

**Comment by mezdahun** [comment]
I think the efficiency will depend on the maximum turning angle the agent can do in a given time step. If you set this higher the agent will react faster to the drop of meter value or increase in meter value. You can imagine this as a physical constraint on the agent's motor system, depending on the nature of the agent this might be slower or faster. I guess also humans in the experimental setup are not able to turn directly backward. 

ezgif com-gif-maker

**Comment by vagechirkov** [comment]
> I think the efficiency will depend on the maximum turning angle the agent can do in a given time step. If you set this higher the agent will react faster to the drop of meter value or increase in meter value. You can imagine this as a physical constraint on the agent's motor system, depending on the nature of the agent this might be slower or faster. I guess also humans in the experimental setup are not able to turn directly backward.


I see, nice! Thank you for the explanation! I think we can leave this parameter as it is (without changing it in the simulation) as it would play a minor role when the resource is relatively small.

**Review by mezdahun — APPROVED** [review]

**Pull request merged**